### PR TITLE
3.2-dev: prettify

### PR DESCRIPTION
**Pull request by Kriechi**
@wilriker as discussed, a few more extensive changes to pydsfapi.

* Adopt the best practices and layout of the PyPi/PyPA sample project.
* minor changes to the API and module layout itself (minimal changes for users required)
* black-ify code for consistent formatting and style.
* improve and fix various typing-related issues with mypy.
* split examples into stand-alone scripts.
* add simple high-level smoke tests based on the example scripts.

Some of the bugfixes I originally planned to send to master are already fixed on the `3.2-dev` branch, so I removed my commits again. 

The commits in this PR should be pretty self-contained and descriptive enough. If they don' apply cleanly on `3.3-dev` please let me know and I can rebase + fix it, once this PR has landed on `3.2-dev`.

### CODE REVIEW

File: pydsfapi/README.md [hunk 1]
```diff
@@ -1,8 +1,11 @@
 # pydsfapi
-Python interface to access DuetSoftwareFramework. It requires at least Python 3.5.
```
Comment (wilriker) [inline pydsfapi/README.md:2]: Why did you remove the notion about Python 3.5 as a minimum?

File: pydsfapi/setup.py [hunk 2]
```diff
@@ -1,18 +1,45 @@
...
-          'pydsfapi.model',
-          ],
-      license='LGPLv3',
-      )
+here = pathlib.Path(__file__).parent.resolve()
+long_description = (here / "README.md").read_text(encoding="utf-8")
+
+setup(
+    name="pydsfapi",
+    version="3.1.3",
```
Comment (wilriker) [inline pydsfapi/setup.py]: Please change the version to 3.2.0

File: pydsfapi/setup.py [hunk 3]
```diff
@@ -1,18 +1,45 @@
...
+        "Programming Language :: Python :: 3.6",
+        "Programming Language :: Python :: 3.7",
+        "Programming Language :: Python :: 3.8",
+        "Programming Language :: Python :: 3.9",
+        "Programming Language :: Python :: 3 :: Only",
+    ],
+    keywords="Duet3D, DuetSoftwareFramework, DSF",
+    package_dir={"": "src"},
+    packages=find_packages(where="src"),
+    python_requires=">=3.6, <4",
```
Comment (wilriker) [inline pydsfapi/setup.py:33]: Last time I checked on DuetPi Python 3.5 was installed (I personally don't use DuetPi, so I do not check very often). So unless this has been updated to at least 3.6 this needs to be compatible to Python 3.5.

File: pydsfapi/src/pydsfapi/__init__.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+SOCKET_DIRECTORY = "/var/run/dsf"
```
Comment (wilriker) [inline pydsfapi/src/pydsfapi/__init__.py]: Please change to `/run/dsf`. Although DSF actually still uses `/var/run/dsf` the folder `/var/run` has been deprecated by systemd and `/run` should be used instead.

File: pydsfapi/src/pydsfapi/commands/basecommands.py [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+
+def cancel():
+    return BaseCommand("Cancel")
+
+
+def ignore():
+    return BaseCommand("Ignore")
+
+
+def get_machine_model():
```
Comment (wilriker) [inline pydsfapi/src/pydsfapi/commands/basecommands.py:88]: Please add Deprecated comment to the user should use the `_object_` function instead. Please do so for all `_machine_` functions.

File: pydsfapi/src/pydsfapi/commands/basecommands.py [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+    endpoint_type: HttpEndpointType, namespace: str, path: str, is_upload_request: bool
+):
+    """
+    Register a new HTTP endpoint via DuetWebServer.
+    This will create a new HTTP endpoint under /machine/{Namespace}/{EndpointPath}.
+    Returns a path to the UNIX socket which DuetWebServer will connect to whenever a matching
+    HTTP request is received. A plugin using this command has to open a new UNIX socket with
+    the given path that DuetWebServer can connect to
+    """
+    if not isinstance(endpoint_type, HttpEndpointType):
```
Comment (wilriker) [inline pydsfapi/src/pydsfapi/commands/basecommands.py:130]: Great work in type checking! :+1:

File: pydsfapi/tox.ini [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+deps =
+    check-manifest
+    black
+    mypy
+    flake8
+    pytest
+commands =
+    check-manifest --ignore 'tox.ini,docs/**,examples/**,tests/**'
+    python setup.py check -m -s
+    black --check --target-version py36 .
```
Comment (wilriker) [inline pydsfapi/tox.ini:14]: Again, if DuetPi is still running 3.5 (I hope it's not) this needs to check for compatibility with 3.5.

**Review by wilriker — CHANGES_REQUESTED** [review]

**Comment by wilriker** [comment]
First of all thanks for the great work. I made some minor change requests, should be pretty quick to fix.

File: pydsfapi/setup.py (hunk 3 above)
Comment (Kriechi) [reply pydsfapi/setup.py:33]: https://duet3d.dozuki.com/Wiki/SBC_Setup_for_Duet_3 links to https://pkg.duet3d.com/DuetPi-lite.zip currently contains a `2021-02-15-DuetPi-lite.img` which ships a `python3.7` executable.

Raspbian Buster comes with python 3.7.
Raspbian Stretch comes with python 3.5 - but I don't think DuetDSF is "that old" to have ever used Stretch?

File: pydsfapi/tox.ini (hunk 7 above)
Comment (Kriechi) [reply pydsfapi/tox.ini:14]: see above discussion - I think we are fine here.
Raspbian Buster comes with python 3.7, and the latest https://pkg.duet3d.com/DuetPi-lite.zip is Buster.

File: pydsfapi/src/pydsfapi/__init__.py (hunk 4 above)
Comment (Kriechi) [reply pydsfapi/src/pydsfapi/__init__.py]: This change is just copied over from another file.
I'll change it - I'm all in favour of adopting the new system layout!

File: pydsfapi/setup.py (hunk 2 above)
Comment (Kriechi) [reply pydsfapi/setup.py]: fixed with latest push.

File: pydsfapi/src/pydsfapi/commands/basecommands.py (hunk 5 above)
Comment (Kriechi) [reply pydsfapi/src/pydsfapi/commands/basecommands.py:88]: Already deprecated them for 3.2.x? Or are you thinking now of the `3.3-dev` branch?

Comment (Kriechi) [reply pydsfapi/src/pydsfapi/commands/basecommands.py:88]: Ah, I think you are looking for the comments in https://github.com/Duet3D/DSF-APIs/pull/4/files#diff-f9a619cbfb505ffc90cb728f7b7e09d502828d2f9fed1e958fcd9fc385f38d0cR245

IMO the high-level functions in the connection classes should be marked as deprecated (which they still are). The underlying `basecommands` should not really be used by users anyway. Happy to mark them as deprecated in a follow-up PR or on 3.3-dev.

File: pydsfapi/setup.py (hunk 3 above)
Comment (wilriker) [reply pydsfapi/setup.py:33]: Latest DuetPi seems to be shipped with 3.7. So everything fine.

File: pydsfapi/src/pydsfapi/commands/basecommands.py (hunk 5 above)
Comment (wilriker) [reply pydsfapi/src/pydsfapi/commands/basecommands.py:88]: No worries, I only later realized that the functions are already marked as Deprecated.
